examples/structured_data/collaborative_filtering_movielens.py

```diff
@@ -0,0 +1,217 @@
+df["movie"] = movie_enc.fit_transform(df["movieId"].values)
+n_movies = df["movie"].nunique()
+df["rating"] = df["rating"].values.astype(np.float32)
+min_rating = min(df["rating"])
+max_rating = max(df["rating"])
+print(
+    "Number of users: {}, Number of Movies: {}, Min rating: {}, Max rating: {}".format(
+        n_users, n_movies, min_rating, max_rating
+    )
+)
+
+"""
+## Prepare training and validation data
+"""
+
+X = df[["user", "movie"]].values
+y = df["rating"].values
+X_train, X_test, y_train, y_test = train_test_split(
+    X, y, test_size=0.1, random_state=42
+)
+print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)
+
+"""
+## Create the model
+
+Let's fix some parameters in the architecture. We will use user embeddings and movie
+embeddings, let's keep the dimensions as 50 for both.
+"""
+EMBEDDING_SIZE = 50
+
+
+class EmbeddingLayer(layers.Layer):
+    def __init__(self, n_items, n_factors):
+        super(EmbeddingLayer, self).__init__()
+        self.n_items = n_items
+        self.n_factors = n_factors
+        self.emb = Embedding(
+            self.n_items,
+            self.n_factors,
+            embeddings_initializer="he_normal",
+            embeddings_regularizer=l2(1e-6),
+        )
+
+    def call(self, x):
+        x = self.emb(x)
+        x = Reshape((self.n_factors,))(x)
+        return x
+
+
+class RecommenderNet(keras.Model):
+    def __init__(self, n_users, n_movies, embedding_size, min_rating, max_rating):
+        super(RecommenderNet, self).__init__(name="recnet")
+        self.n_users = n_users
+        self.n_movies = n_movies
+        self.embedding_size = embedding_size
+        self.min_rating = min_rating
+        self.max_rating = max_rating
+        self.user_embedding = Embedding(
+            n_users,
+            embedding_size,
+            embeddings_initializer="he_normal",
+            embeddings_regularizer=l2(1e-6),
+        )
+        self.user_bias = Embedding(n_users, 1)
+        self.movie_embedding = Embedding(
+            n_movies,
+            embedding_size,
+            embeddings_initializer="he_normal",
+            embeddings_regularizer=l2(1e-6),
+        )
+        self.movie_bias = Embedding(n_movies, 1)
+
+    def call(self, inputs):
+        u = self.user_embedding(inputs[:, 0])
+        ub = self.user_bias(inputs[:, 0])
+        m = self.movie_embedding(inputs[:, 1])
+        mb = self.movie_bias(inputs[:, 1])
+        x = Dot(axes=1)([u, m])
+        x = Add()([x, ub, mb])
+        x = Activation("sigmoid")(x)
+        x = Lambda(lambda x: x * (self.max_rating - self.min_rating) + self.min_rating)(
+            x
+        )
+        return x
+
+
+inputs = layers.Input(shape=(2,))
+recommendation_layer = RecommenderNet(
+    n_users, n_movies, EMBEDDING_SIZE, min_rating, max_rating
+)
+outputs = recommendation_layer(inputs)
+model = keras.Model(inputs=inputs, outputs=outputs)
+opt = Adam(lr=0.001)
+model.compile(loss="mean_squared_error", optimizer=opt)
```

Comment:
Why use MSE instead of binary crossentropy (assuming that you normalize the movie ratings between 0 and 1)?

Reply:
The scores are between 0.5 and 5 (minimum and maximum ratings, respectively). I use a sigmoid layer to get a number between 0 and 1, and then scale it back using the minimum and maximum ratings.

Reply:
Try it out -- regression done with NNs often works better with crossentropy.

Reply:
Thanks @fchollet for the recommendation. I tried it and looks good. Applied scaling to the dataset. Updated code has been committed.